src/compile.ts

```diff
@@ -856,6 +1175,167 @@ export function compile(
         return;
       }
 
+      /**
+       * Attempts to find the a version of a reference that is outside of a certain scope.
+       *
+       * This is useful for finding variables that have been instantiated outside of a closure, but
+       * renamed inside of the closure.
+       *
+       * When serializing the lambda functions, we want references from outside of the closure if possible.
+       *
+       * ```ts
+       * const bus = new EventBus()
+       * new Function(() => {
+       *     const busbus = bus;
+       *     busbus(...)
+       * })
+       * ```
+       *
+       * Can also follow property access.
+       *
+       * ```ts
+       * const x = { y : () => {} };
+       *
+       * () => {
+       *    const z = x;
+       *    z.y() // x.y() is returned
+       * }
+       * ```
+       *
+       * getOutOfScopeValueNode(z.y) => x.y
+       *
+       * ```ts
+       * const x = () => {};
+       *
+       * () => {
+       *    const z = { y: x };
+       *    z.y()
+       * }
+       * ```
+       *
+       * getOutOfScopeValueNode(z.y) => x
+       *
+       * The call to busbus can be resolved to bus if the scope is the array function.
+       */
+      function getOutOfScopeValueNode(
```

Comment:
Probably cannot be found. Have not exhaustively tested external reference patterns. Good test to add.

Reply:
Let's open an issue or track in a test?

Reply:
got it :-) (commit soon)

```ts
const b = {
    bus: new EventBus(...)
}
new Function(... () => {
    const { bus } = b;
    bus(...);
})
```

```ts
{ integration: b["bus"] }
```

Reply:
summary: created a method to discover and flatten destructured assignment.

```ts
      function flattenDestructuredAssignment(
        element: ts.BindingElement
      ): ts.ElementAccessExpression {
        // if the binding renames the property, get the original
        // { a : x } -> a
        // { a } -> a
        // [a] -> 0
        const name = ts.isArrayBindingPattern(element.parent)
          ? element.pos
          : // binding renames the property or is a nested binding pattern.
          element.propertyName
          ? element.propertyName
          : // the "name" can be a binding pattern. In that case the propertyName will be set.
            (element.name as ts.Identifier);

        const getParent = () => {
          // { a } = b;
          if (ts.isVariableDeclaration(element.parent.parent)) {
            if (!element.parent.parent.initializer) {
              throw Error(
                "Expected a initializer on a destructured assignment: " +
                  element.getText()
              );
            }
            return element.parent.parent.initializer;
          } else if (ts.isBindingElement(element.parent.parent)) {
            return flattenDestructuredAssignment(element.parent.parent);
          } else {
            throw Error(
              "Cannot flatten destructured parameter: " + element.getText()
            );
          }
        };

        const parent = getParent();

        // always use element access as this will work for all possible values.
        // [parent][name]
        return ts.factory.createElementAccessExpression(
          parent,
          typeof name !== "number"
            ? ts.isComputedPropertyName(name)
              ? name.expression
              : ts.isIdentifier(name)
              ? // { a } = b -> b["a"]
                ts.factory.createStringLiteral(name.text)
              : name
            : // numeric index position for a destructured array
              name
        );
      }
```

Reply:
So the synth-time interpreters will never have to handle restructuring? Is there any chance of name conflicts after flattening?

Reply:
We are not always flattening. I am only doing it for getting the "out of scope" version of the integration to be able to pass back into the function. In this case at least, the names won't collide.